pitaya-sharp/NPitaya/src/Metrics/MetricsReporter.cs

```diff
@@ -0,0 +1,103 @@
+using System;
+using System.Runtime.InteropServices;
+using NPitaya.Models;
+
+namespace NPitaya.Metrics
+{
+    public class MetricsReporter
+    {
+        private const string LabelSeparator = "_";
+        private const string PitayaSubsystem = "pitaya";
+
+        public IntPtr Ptr { get; }
+
+        public MetricsReporter(PrometheusReporter prometheusReporter)
+        {
+            var handle = GCHandle.Alloc(prometheusReporter, GCHandleType.Normal);
+            var reporterPtr = GCHandle.ToIntPtr(handle);
+            Ptr = PitayaCluster.pitaya_metrics_reporter_new(
+                RegisterCounterFn,
+                RegisterHistogramFn,
+                RegisterGaugeFn,
+                IncCounterFn,
+                ObserveHistFn,
+                SetGaugeFn,
+                AddGaugeFn,
+                reporterPtr);
+        }
+
+        ~MetricsReporter()
+        {
+            PitayaCluster.pitaya_metrics_reporter_drop(Ptr);
+        }
+
+        static void RegisterCounterFn(IntPtr prometheusPtr, MetricsOpts opts)
+        {
+            var ns = Marshal.PtrToStringAnsi(opts.MetricNamespace);
+            // TODO (felipe.rodopoulos): prometheus-net does not support subsystem label yet. We'll use a hardcoded one.
+            var name = Marshal.PtrToStringAnsi(opts.Name);
+            var prometheusReporter = RetrievePrometheus(prometheusPtr);
+            var key = BuildKey(name);
+            prometheusReporter.RegisterCounter(key);
+        }
+
+        static void RegisterHistogramFn(IntPtr prometheusPtr, MetricsOpts opts)
+        {
+            var ns = Marshal.PtrToStringAnsi(opts.MetricNamespace);
+            // TODO (felipe.rodopoulos): prometheus-net does not support subsystem label yet. We'll use a hardcoded one.
+            var name = Marshal.PtrToStringAnsi(opts.Name);
+            var prometheusReporter = RetrievePrometheus(prometheusPtr);
+            var key = BuildKey(name);
+            prometheusReporter.RegisterHistogram(key);
```

Comment:
Are you not implementing labels yet?

Reply:
Not yet, I'm leaving it to a next step, because I'm concerned with language and system metrics first. They are more important and we are not using labels yet.

Reply:
Cool